cvat/apps/engine/serializers.py

```diff
@@ -3499,61 +3771,74 @@ class Meta:
+
+                qs_with_similar_ones = models.CloudStorage.objects.filter(
+                    provider_type=original_cs.provider_type,
+                    resource=original_cs.resource,
+                    organization_id=dst_organization_id,
+                ).annotate(Count("manifests")).filter(manifests__count=original_cs.manifests__count)
+
+                for cs in qs_with_similar_ones:
+                    if (
+                        cs.get_specific_attributes() != original_cs.get_specific_attributes()
+                        # we do not need to compare manifest contents since the content
+                        # will be synchronized with the original CS file on the next content request
+                        or list(cs.manifests.values_list("filename", flat=True).order_by("filename"))
+                        != list(original_cs.manifests.values_list("filename", flat=True).order_by("filename"))
+                    ):
+                        continue
 
-        # storage_instance maybe None
-        storage_instance = getattr(instance, storage_type)
+                    similar_ones.append(cs)
 
-        if new_cloud_storage_id:
-            if new_location and new_location != models.Location.CLOUD_STORAGE:
-                raise serializers.ValidationError(
-                    f"It is not allowed to specify '{new_location}' location together with cloud storage id"
-                )
-            elif (
-                not new_location
-                and getattr(storage_instance, "location", None) != models.Location.CLOUD_STORAGE
-            ):
-                raise serializers.ValidationError(
-                    f"The configuration of {storage_type} is not full"
-                )
+                if len(similar_ones) != 1:
+                    raise serializers.ValidationError(msg_no_similar_cs)
 
-            if not models.CloudStorage.objects.filter(id=new_cloud_storage_id).exists():
-                raise serializers.ValidationError(
-                    f"The specified cloud storage {new_cloud_storage_id} does not exist."
-                )
-        else:
-            if new_location == models.Location.CLOUD_STORAGE:
-                raise serializers.ValidationError(
-                    "Cloud storage was selected as location but its id was not specified"
-                )
-            elif (
-                not new_location
-                and getattr(storage_instance, "location", None) == models.Location.CLOUD_STORAGE
-                and "cloud_storage_id" in new_conf
-            ):
-                raise serializers.ValidationError(
-                    "It is not allowed to reset a cloud storage id without explicitly resetting a location"
-                )
+                new_conf = {
+                    "location": models.Location.CLOUD_STORAGE,
+                    "cloud_storage_id": similar_ones[0].pk
+                }
```

Comment:
Consider moving everything after `get()` into the `else` block of try-except-else, the try section seems to be too broad for the exception being catched.